packages/labs/ssr/src/lib/render-value.ts

```diff
@@ -83,7 +87,10 @@ const patchedDirectiveCache = new WeakMap<DirectiveClass, DirectiveClass>();
 const patchIfDirective = (value: unknown) => {
   // This property needs to remain unminified.
   const directiveCtor = getDirectiveClass(value);
-  if (directiveCtor !== undefined) {
+  if (
+    directiveCtor !== undefined &&
+    (directiveCtor as MaybePatchedDirective).patchedDirective === undefined
```

Comment:
Why can't this just check the `patchedDirectiveCache`?

```suggestion
!patchedDirectiveCache.has(directiveCtor)
```

Otherwise consider a WeakSet?

Reply:
It's the first check that happens within this if block, but the problem was that `directiveCtor` could be an already patched one which isn't added as the cache map key.

after patching the directive we do
```ts
patchedDirectiveCache.set(directiveCtor, patchedCtor);
```
so only the unpatched `directiveCtor` serves as the key.

I entertained the idea of also adding
```ts
patchedDirectiveCache.set(patchedCtor, patchedCtor);
```
which would make lookup from the cache catch already patched ones, but wasn't sure if that's really better than a sentinel property on the patched ctor.

Reply:
hmm... I think we need some more comments here and make a name change? I would have thought that patching a directive would have actually patched the directive ctor we received, so that there isn't a _new_ class. We're instead _replacing_ the directive.

I am curious how a replaced directive re-enters here. When is a DirectiveResult passed to multiple renders?

Regardless, I get the problem now, and I think there may be a much simpler solution to this whole thing, which is to monkey-patch `_$resolve` on `Directive`. That method is used by all directives. We just have to check that we patched `Directive` in case we get multiple copies of lit-html in play.

Reply:
Great discussion. I'll explore a simpler fix.

Reply:
Previously this was implemented as a patch to the Directive prototype: https://github.com/lit/lit/pull/1599/commits/73930f40ac1d1033eed205d723498f6792a4ddba#diff-b8cfd0fdb8410f04f051c052e6ce548e5403f407d6954d48df02b7c5bdb0e506R75-R85

And `_$resolve` was patched directly. See change here: https://github.com/lit/lit/pull/1599/commits/96dfb42234a96782ceab5b3cfa091310020b4927#diff-f57f5f57247e8497a15da483cc93126465912bec0dd886454289befb86a72cbdL36-L41

History of this change is captured in the PR: https://github.com/lit/lit/pull/1599

Reply:
> I am curious how a replaced directive re-enters here. When is a DirectiveResult passed to multiple renders?

Hmm... the DirectiveResult would be in the values array of a template, and we mutate that value, so I think if that same template is getting SSRed over multiple requests, this could happen?

> monkey-patch `_$resolve` on `Directive`

Catching all copies might be hard.. we can only dig into the prototype of the directive class we got passed in. and we need this baked into `lit-html/private-ssr-support.js` due to the stable property mangling.

Reply:
I think there's another way to work around the issue of multiple lit-html libraries:

```ts
function ssrResolve(part: Part, props: Array<unknown>): unknown {
  return this.render(props);
}

const patchDirectiveResolve = (value: unknown) => {
  const directiveCtor = getDirectiveClass(value);
  if (directiveCtor._$resolve !== ssrResolve) {
    // even better, walk up the proto chain to Directive
    directiveCtor._$resolve = ssrResolve;
  }
};
```

This will work against any compatible version of the lit-html library.

Reply:
Done!